server/src/main/java/org/elasticsearch/search/aggregations/metrics/GeoCentroidAggregator.java

```diff
@@ -70,42 +76,65 @@ public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCol
+                weightCompensations = bigArrays.grow(weightCompensations, bucket + 1);
                 counts = bigArrays.grow(counts, bucket + 1);
+                dimensionalShapeTypes = bigArrays.grow(dimensionalShapeTypes, bucket + 1);
 
                 if (values.advanceExact(doc)) {
                     final int valueCount = values.docValueCount();
                     // increment by the number of points for this document
                     counts.increment(bucket, valueCount);
                     // Compute the sum of double values with Kahan summation algorithm which is more
                     // accurate than naive summation.
+                    DimensionalShapeType shapeType = DimensionalShapeType.fromOrdinalByte(dimensionalShapeTypes.get(bucket));
                     double sumLat = latSum.get(bucket);
                     double compensationLat = latCompensations.get(bucket);
                     double sumLon = lonSum.get(bucket);
                     double compensationLon = lonCompensations.get(bucket);
+                    double sumWeight = weightSum.get(bucket);
+                    double compensatedWeight = weightCompensations.get(bucket);
 
                     compensatedSumLat.reset(sumLat, compensationLat);
                     compensatedSumLon.reset(sumLon, compensationLon);
+                    compensatedSumWeight.reset(sumWeight, compensatedWeight);
 
                     // update the sum
                     for (int i = 0; i < valueCount; ++i) {
                         MultiGeoValues.GeoValue value = values.nextValue();
-                        //latitude
-                        compensatedSumLat.add(value.lat());
-                        //longitude
-                        compensatedSumLon.add(value.lon());
+                        int compares = DimensionalShapeType.COMPARATOR.compare(shapeType, value.dimensionalShapeType());
+                        if (compares < 0) {
+                            double coordinateWeight = value.weight();
+                            compensatedSumLat.reset(coordinateWeight * value.lat(), 0.0);
+                            compensatedSumLon.reset(coordinateWeight * value.lon(), 0.0);
+                            compensatedSumWeight.reset(coordinateWeight, 0.0);
+                            dimensionalShapeTypes.set(bucket, (byte) value.dimensionalShapeType().ordinal());
+                        } else if (compares == 0) {
+                            double coordinateWeight = value.weight();
+                            // weighted latitude
+                            compensatedSumLat.add(coordinateWeight * value.lat());
+                            // weighted longitude
+                            compensatedSumLon.add(coordinateWeight * value.lon());
+                            // weight
+                            compensatedSumWeight.add(coordinateWeight);
+                        }
```

Comment:
are we missing here the cat when. compare > 0?

Reply:
if the running max-shapetype is greater than the current document's shape-type, then the centroid of the document is not considered. I will add more comments here

Reply:
you are right, a small comment would be great